p-token/test-properties/thunks.md

```diff
@@ -0,0 +1,263 @@
+| test_process_transfer                         | 2                            | [2](#2) [3](#3)                         |
+| test_process_amount_to_ui_amount              | 2                            | [2](#2) [7](#7)                         |
+| test_process_ui_amount_to_amount              | stuck                        | stuck                                   |
+
+
+
+## 1
+
+### Full thunk:
+
+```
+thunk ( #cast ( Range ( ListItem (Integer ( ARG_UINT214:Int , 8 , false ))
+                        ListItem (Integer ( ARG_UINT215:Int , 8 , false ))
+                        ListItem (Integer ( ARG_UINT216:Int , 8 , false ))
+                        ListItem (Integer ( ARG_UINT217:Int , 8 , false ))
+                        ListItem (Integer ( ARG_UINT218:Int , 8 , false ))
+                        ListItem (Integer ( ARG_UINT219:Int , 8 , false ))
+                        ListItem (Integer ( ARG_UINT220:Int , 8 , false ))
+                        ListItem (Integer ( ARG_UINT221:Int , 8 , false ))
+                        ) , castKindTransmute , ty ( 600721 ) , ty ( 600142 ) ) )
+```
+
+### Explanation:
+
+Cast between an array of 8 `u8`s (`ty ( 600721 )`) and a `U64` (`ty ( 600142 )`).
+
+### Safety
+
+??
+
+### Appearances
+
+Appears on:
+- `test_process_approve_checked`
+- `test_process_transfer`
+
+## 2
+
+### Full thunk
+
+```
+ thunk ( operandConstant ( constOperand ( ... span: span ( 600186 ) , userTy: noUserTypeAnnotationIndex ,
+    const: mirConst ( ... kind: constantKindZeroSized , ty: ty ( 600113 ) , id: mirConstId ( 375 ) ) ) ) )
+```
+
+### Explanation:
+
+??
+
+### Safety
+
+??
+
+### Appearances
+
+Appears on:
+- `test_process_approve`
+- `test_process_approve_checked`
+- `test_process_withdraw_excess_lamports_account`
+- `test_process_transfer`
+
+## 3
+
+### Full thunk
+
+```
+thunk ( #cast ( Integer ( ?STATE:Int , 8 , false ) , castKindTransmute , ty ( 600077 ) , ty ( 600110 ) ) )
+```
+
+### Explanation:
+
+The thunked cast is a transmute between `u8` (`ty ( 600077 )`) and the enum `AccountState` (`ty ( 600110 )`),
+which has three variants: `Uninitialized`, `Initialized` and `Frozen`.
+
+### Safety:
+
+The rule that seems to make this work is the following:
+
+```
+  rule <k> #discriminant(
+              thunk(#cast (Integer(DATA, _, false), castKindTransmute, _, TY)),
+              TY
+            ) => Integer(DATA, 0, false) // HACK: bit width 0 means "flexible"
+          ...
+        </k>
+    requires #isEnumWithoutFields(lookupTy(TY))
+```
+
+The rule is not taking into account the width of the transmuted integer.
+
+### Appearances
+
+Appears on:
+- `test_process_approve`
+- `test_process_approve_checked`
+- `test_process_withdraw_excess_lamports_account`
+
+## 4
+
+### Full Thunk
+
+```
+thunk ( #cast ( Float ( 0.20000000000000000e1 , 64 ) , castKindTransmute , ty ( 601188 ) , ty ( 600142 ) ) )
```

Comment:
We already have [a specific rule which compares this against the expected `u64` with the same bit pattern](https://github.com/runtimeverification/mir-semantics/blob/26807db3fcc3089df82b5196095b8e41890fafc4/kmir/src/kmir/kdist/mir-semantics/symbolic/p-token.md#rent-specific-evaluation-for-float-exempt-threshold).
If we don't want to have the `thunk` at all, we can rewrite this `#cast` with this specific `Float` value to the expected `u64`.